File: packages/node/lib/tools/terminal.ts
import { ClientReadableStream } from '@grpc/grpc-js';
import { ProcessEvent, ProcessInfo } from '../generated/terminal';
import { TerminalService } from '../services/terminal';
import { CommandResult, ConnectToProcessOptions, RunCommandOptions } from '../types/terminal';

export class TerminalTool {
  private readonly terminalService: TerminalService;

  constructor(template: string, instanceId: string) {
    this.terminalService = new TerminalService(template, instanceId);
  }

  /**
   * Runs a command and waits until it finishes executing
   * @param cmd Command to execute
   * @param options Options for running the command
   * @returns Command result
   */
  async run(cmd: string, options?: RunCommandOptions & { background?: false }): Promise<CommandResult>;

  /**
   * Runs a command in the background
   * @param cmd Command to execute
   * @param options Options for running the command
   * @returns Process handle
   */
  async run(cmd: string, options?: RunCommandOptions & { background?: true }): Promise<ProcessHandle>;

  // Implementation for all overloads
  async run(cmd: string, options?: RunCommandOptions) {
    const stream = this.terminalService.spawnProcess({
      cmd: '/bin/bash',
      args: ['-l', '-c', cmd],
      cwd: options?.cwd || '/',
      env: options?.env || {},
      cols: options?.cols || 80,
      rows: options?.rows || 24
    });

    const pid = await TerminalTool.waitForPid(stream);

    const handle = new ProcessHandle(
      pid,
      stream,
      this.terminalService,
      options?.onData,
      options?.onExit,
      options?.onError,
      options?.timeoutMs
    );

    if (!options?.background) {
      return handle.wait();
    }

    return handle;
  }

  /**
   * List all running processes
   * @returns Array of process information
   */
  async listProcesses(): Promise<ProcessInfo[]> {
    const response = await this.terminalService.listProcesses({});
    return response.processes;
  }

  /**
   * Connect to a running process
   * @param pid Process ID to connect to
   * @returns Process handle
   */
  connect(pid: number, options?: ConnectToProcessOptions): ProcessHandle {
    const stream = this.terminalService.connectToProcess({ pid });
    return new ProcessHandle(
      pid,
      stream,
      this.terminalService,
      options?.onData,
      options?.onExit,
      options?.onError,
      options?.timeoutMs
    );
  }

  private static waitForPid(stream: ClientReadableStream<ProcessEvent>): Promise<number> {
    return new Promise((resolve, reject) => {
      stream.on('data', (event: ProcessEvent) => {
        if (event.type === 'START') {
          resolve(event.pid);
        } else if (event.type === 'ERROR') {
          reject(new Error(event.error));
        } else if (event.type === 'EXIT') {
          reject(new Error('Process killed before starting'));
        }
      });
      stream.on('error', (err) => {
        reject(err);
      });
      stream.on('end', () => {
        reject(new Error('Process killed before starting'));
      });
    });
  }
}

export class ProcessHandle {
  private _data: string;
  private _result: CommandResult | null;
  private _wait: Promise<void>;
  private _timeoutId: NodeJS.Timeout | null;
  private readonly decoder: TextDecoder;

  constructor(
    readonly pid: number,
    private readonly stream: ClientReadableStream<ProcessEvent>,
    private readonly terminalService: TerminalService,
    private readonly onData?: (data: string) => void | Promise<void>,
    private readonly onExit?: (exitCode: number) => void | Promise<void>,
    private readonly onError?: (error: Error) => void | Promise<void>,
    private readonly timeoutMs?: number,
    private readonly encoding: BufferEncoding = 'utf-8'
  ) {
    this._data = '';
    this._result = null;
    this._timeoutId = null;
    this.decoder = new TextDecoder(encoding, { fatal: false });
    this._wait = this.handleStream();
  }

  get data(): string {
    return this._data;
  }

  get exitCode(): number | null {
    return this._result ? this._result.exitCode : null;
  }

  private clearTimeout() {
    if (this._timeoutId) {
      clearTimeout(this._timeoutId);
      this._timeoutId = null;
    }
  }

  private handleStream(): Promise<void> {
    return new Promise((resolve, reject) => {
      if (this.timeoutMs) {
        this._timeoutId = setTimeout(() => {
          const error = new Error(`Command timed out after ${this.timeoutMs}ms`);
          if (this.onError) {
            this.onError(error);
          }
          reject(error);
          this.kill('SIGTERM');
        }, this.timeoutMs);
      }

      this.stream.on('data', (event: ProcessEvent) => {
        if (event.type === 'DATA') {
          const data = this.decoder.decode(event.data);
          this._data += data;
          if (this.onData) {
            this.onData(data);
          }
        } else if (event.type === 'EXIT') {
          this.clearTimeout();
          this._result = {
            exitCode: event.exitCode,
            data: this._data
          };
          if (this.onExit) {
            this.onExit(event.exitCode);
          }
          resolve();
        } else if (event.type === 'ERROR') {
          this.clearTimeout();
          const error = new Error(event.error);
          if (this.onError) {
            this.onError(error);
          }
          reject(error);
        }
      });
      this.stream.on('error', (err) => {
        this.clearTimeout();
        if (this.onError) {
          this.onError(err);
        }
        reject(err);
      });
      this.stream.on('end', () => {
        this.clearTimeout();
        if (this._result) {
          if (this.onExit) {
            this.onExit(this._result.exitCode);
          }
          resolve();
        } else {
          const err = new Error('Process exited without a result');
          if (this.onError) {
            this.onError(err);
          }
          reject(err);
        }
      });
    });
  }

  /**
   * Wait for the process to finish executing
   * @returns Command result
   */
  public async wait(): Promise<CommandResult> {
    await this._wait;
    if (this._result) {
      return this._result;
    }
    throw new Error('Process exited without a result');
  }

  /**
   * Write data to the command's process
   * @param data Data to write
   */
  public async write(data: string | Buffer): Promise<void> {
    const buffer = Buffer.isBuffer(data) ? data : Buffer.from(data, this.encoding);
    await this.terminalService.sendInput({ pid: this.pid, input: buffer });
  }

  /**
   * Resize the process's terminal
   * @param cols Number of columns
   * @param rows Number of rows
   */
  public async resize(cols: number, rows: number): Promise<void> {
    await this.terminalService.updateProcess({ pid: this.pid, cols, rows });
  }

  /**
   * Kill the process
   * @param signal Signal to send to the process
   */
  public async kill(signal = 'SIGTERM'): Promise<void> {
    await this.terminalService.sendSignal({ pid: this.pid, signal });
  }
}

File: packages/python/metallic/tools/terminal.py
"""Terminal tool for process management."""
import asyncio
from typing import List, Dict, Any, Optional, AsyncIterator

from ..services.terminal import TerminalService
from ..generated import terminal_pb2


class ProcessInfo:
    """Process information data structure."""
    
    def __init__(self, pid: int, command: str, cwd: str, cols: int, rows: int, start_time: int):
        self.pid = pid
        self.command = command
        self.cwd = cwd
        self.cols = cols
        self.rows = rows
        self.start_time = start_time


class ProcessEvent:
    """Process event data structure."""
    
    def __init__(self, pid: int, event_type: str, data: bytes = b"", error: str = "", exit_code: int = 0):
        self.pid = pid
        self.type = event_type
        self.data = data
        self.error = error
        self.exit_code = exit_code


class ProcessHandle:
    """Handle for managing a spawned process."""
    
    def __init__(self, pid: int, terminal_service: TerminalService):
        self.pid = pid
        self._terminal_service = terminal_service
    
    async def send_input(self, input_data: bytes) -> None:
        """Send input to the process.
        
        Args:
            input_data: Input data to send
        """
        request = terminal_pb2.SendInputRequest(pid=self.pid, input=input_data)
        response = await self._terminal_service.send_input(request)
        
        if not response.success:
            raise Exception(response.error)
    
    async def update_size(self, cols: int, rows: int) -> None:
        """Update the terminal size.
        
        Args:
            cols: Number of columns
            rows: Number of rows
        """
        request = terminal_pb2.UpdateProcessRequest(pid=self.pid, cols=cols, rows=rows)
        response = await self._terminal_service.update_process(request)
        
        if not response.success:
            raise Exception(response.error)
    
    async def send_signal(self, signal: str) -> None:
        """Send a signal to the process.
        
        Args:
            signal: Signal name (e.g., "SIGTERM", "SIGKILL")
        """
        request = terminal_pb2.SendSignalRequest(pid=self.pid, signal=signal)
        response = await self._terminal_service.send_signal(request)
        
        if not response.success:
            raise Exception(response.error)


class TerminalTool:
    """Tool for terminal and process management."""
    
    def __init__(self, project_id: str, instance_id: str):
        """Initialize the Terminal tool.
        
        Args:
            project_id: Project ID
            instance_id: Instance ID
        """
        self.terminal = TerminalService(project_id, instance_id)
    
    async def close(self):
        """Close the terminal service connection."""
        await self.terminal.close()
    
    async def list_processes(self) -> List[ProcessInfo]:
        """List running processes.
        
        Returns:
            List of ProcessInfo objects
        """
        request = terminal_pb2.ListProcessesRequest()
        response = await self.terminal.list_processes(request)
        
        if not response.success:
            raise Exception(response.error)
        
        return [
            ProcessInfo(
                pid=proc.pid,
                command=proc.command,
                cwd=proc.cwd,
                cols=proc.cols,
                rows=proc.rows,
                start_time=proc.startTime
            )
            for proc in response.processes
        ]
    
    async def spawn_process(
        self,
        cmd: str,
        cwd: str = "",
        env: Optional[Dict[str, str]] = None,
        args: Optional[List[str]] = None,
        cols: int = 80,
        rows: int = 24
    ) -> AsyncIterator[ProcessEvent]:
        """Spawn a new process.
        
        Args:
            cmd: Command to run
            cwd: Working directory
            env: Environment variables
            args: Command arguments
            cols: Terminal columns
            rows: Terminal rows
            
        Yields:
            ProcessEvent objects
        """
        request = terminal_pb2.SpawnProcessRequest(
            cmd=cmd,
            cwd=cwd,
            env=env or {},
            args=args or [],
            cols=cols,
            rows=rows
        )
        
        async for event in self.terminal.spawn_process(request):
            if event.error:
                raise Exception(event.error)
            
            yield ProcessEvent(
                pid=event.pid,
                event_type=event.type,
                data=event.data,
                error=event.error,
                exit_code=event.exitCode
            )
    
    async def connect_to_process(self, pid: int) -> AsyncIterator[ProcessEvent]:
        """Connect to an existing process.
        
        Args:
            pid: Process ID to connect to
            
        Yields:
            ProcessEvent objects
        """
        request = terminal_pb2.ConnectToProcessRequest(pid=pid)
        
        async for event in self.terminal.connect_to_process(request):
            if event.error:
                raise Exception(event.error)
            
            yield ProcessEvent(
                pid=event.pid,
                event_type=event.type,
                data=event.data,
                error=event.error,
                exit_code=event.exitCode
            )
    
    async def run(
        self,
        cmd: str,
        cwd: str = "",
        env: Optional[Dict[str, str]] = None,
        args: Optional[List[str]] = None,
        cols: int = 80,
        rows: int = 24
    ) -> ProcessHandle:
        """Run a command and return a process handle.
        
        Args:
            cmd: Command to run
            cwd: Working directory
            env: Environment variables
            args: Command arguments
            cols: Terminal columns
            rows: Terminal rows
            
        Returns:
            ProcessHandle for the spawned process
        """
        pid = None
        async for event in self.spawn_process(cmd, cwd, env, args, cols, rows):
            if event.type == "spawn" and event.pid:
                pid = event.pid
                break
        
        if pid is None:
            raise Exception("Failed to spawn process")
        
        return ProcessHandle(pid, self.terminal)
